test/functional/core/exit_spec.lua

```diff
@@ -0,0 +1,46 @@
+local helpers = require('test.functional.helpers')(after_each)
+
+local command = helpers.command
+local eval = helpers.eval
+local eq = helpers.eq
+local run = helpers.run
+
+describe('v:exiting', function()
+  local cid
+
+  before_each(function()
+    helpers.clear()
+    cid = helpers.nvim('get_api_info')[1]
+  end)
+
+  it('defaults to v:null', function()
+    assert(eval('v:exiting is v:null'))
+  end)
+
+  it('is 0 on normal exit', function()
+    local function on_setup()
+      command('autocmd VimLeavePre * call rpcrequest('..cid..', "")')
+      command('autocmd VimLeave    * call rpcrequest('..cid..', "")')
```

Comment:
I would suggest omitting the VimLeave autocmd, it makes the test unclear. Because it never gets run. Maybe an additional test for VimLeave also would be worth having.